pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -177,18 +177,41 @@ func (c *clusterMemberRemovalController) attemptToScaleDown(ctx context.Context,
 	if err != nil {
 		return fmt.Errorf("could not get a list of unhealthy members: %v", err)
 	}
+
+	// map machines pending deletion to node internal ip
+	votingMachinesPendingDeletionIndex := ceohelpers.IndexMachinesByNodeInternalIP(votingMachinesPendingDeletion)
+	// find unhealthy machine pending deletion to remove first from the cluster, without violating quorum
+	var unhealthyMachinePendingDeletion []*machinev1beta1.Machine
 	if len(unhealthyMembers) > 0 {
 		var unhealthyMembersURLs []string
 		for _, unhealthyMember := range unhealthyMembers {
+			unhealthyMemberNodeInternalIP, err := ceohelpers.MemberToNodeInternalIP(unhealthyMember)
+			if err != nil {
+				return fmt.Errorf("cannot get node internal ip for unhealthy member %v: %w", unhealthyMember, err)
+			}
+			unhealthyMemberMachine, ok := votingMachinesPendingDeletionIndex[unhealthyMemberNodeInternalIP]
+			if ok {
+				unhealthyMachinePendingDeletion = append(unhealthyMachinePendingDeletion, unhealthyMemberMachine)
+			}
 			if len(unhealthyMember.PeerURLs) > 0 {
 				unhealthyMembersURLs = append(unhealthyMembersURLs, unhealthyMember.PeerURLs[0])
 			} else {
 				unhealthyMembersURLs = append(unhealthyMembersURLs, unhealthyMember.Name)
 			}
 		}
-		return fmt.Errorf("cannot proceed wth scaling down, unhealthy members found: %v", unhealthyMembersURLs)
+		if len(unhealthyMachinePendingDeletion) > 0 {
+			klog.V(4).Infof("found unhealthy voting members with machine pending deletion: %v", unhealthyMachinePendingDeletion)
+			klog.V(4).Infof("unhealthy members found: %v", unhealthyMembersURLs)
+		} else {
+			return fmt.Errorf("cannot proceed with scaling down, unhealthy members found: %v", unhealthyMembersURLs)
+		}
 	}
 
+	// remove the unhealthy machine pending deletion first
+	// if no unhealthy machine pending deletion found, then attempt to scale down the healthy machines pending deletion
+	if len(unhealthyMachinePendingDeletion) > 0 {
+		votingMachinesPendingDeletion = append(unhealthyMachinePendingDeletion, votingMachinesPendingDeletion...)
```

Comment:
Resurrecting the earlier [thread](https://github.com/openshift/cluster-etcd-operator/pull/937) on this here: 
If we have 2/4 unhealthy members, we're appending both of them for deletion.
While I can see that in practice we'll break after only removing one, from a code readability perspective it's confusing to have the intent to remove multiple unhealthy members since that could cause quourm loss(although we already have quorum loss at 2/4 anyway).

Also @Elbehery for your earlier comment:
>shouldn't we better fallback to the error when there are too many unhealthy members?

> well, I discussed this with @hasbro17 yesterday, we concluded that as long as the current number of voting members is greater than the controlPlaneReplicaCount , then we can remove members whose machine is pending deletions

I probably mentioned this but the key not just voting members but `healthy voting members == controlPlaneReplicaCount` which would be 3/4 healthy.
And even if it's safe to remove I would err on the side of caution when we have multiple unhealthy members.

So I think we should err out in that case and only address the single member unhealthy case here.

Reply:
Thanks @hasbro17  

> I probably mentioned this but the key not just voting members but healthy voting members == controlPlaneReplicaCount which would be 3/4 healthy.

So [here](https://github.com/openshift/cluster-etcd-operator/blob/17238fb25a3e14d0ef123f26e4070a2673315eb8/pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go#L166) we check that the number of `liveVotingMembers <= controlPlaneReplicaCount`. if the invariant does not hold, we skip scaling down, regardless of the member health. 

We can replace `c.etcdClient.MemberList(ctx)` with `c.etcdClient.HealthyMembers(ctx)` and then we sure that we hold the invariant before proceeding with scaling down, wdyt ? 

> And even if it's safe to remove I would err on the side of caution when we have multiple unhealthy members.

So here I am logging the unhealthy voting members && unhealthy voting members pending deletion. I can error on the case of multiple unhealthy members, but then we prevent scaling down at all, please correct me if i am wrong 

I will create a separate commit with the possible fix to continue the discussion.

Reply:
@hasbro17  I have made the changes in this [commit](https://github.com/openshift/cluster-etcd-operator/pull/937/commits/3968a8c50ba3a049e6310694fdc29aa4a046ad51)

> I probably mentioned this but the key not just voting members but healthy voting members == controlPlaneReplicaCount which would be 3/4 healthy.

so basically I have restricted the invariant of scaling down, considering only `healthy voting members < controlPlaneReplicaCount` breaks the invariant and therefore we skip scaling down regardless

Reply:
> And even if it's safe to remove I would err on the side of caution when we have multiple unhealthy members.

I added this [test case](https://github.com/openshift/cluster-etcd-operator/blob/2d5cb605ec80aebe86cd95426a801bfa2f2e6d79/pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller_test.go#L780) to show that we can allow scaling down having multiple unhealthy machines without violating quorum. If you still wanna to error if we have more than one unhealthy machine pending deletion, ok I change it .. This is just a demonstration :)